File: dbm-ui/backend/flow/engine/controller/spider.py
"""
TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at https://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""

from backend.db_meta.enums import ClusterType
from backend.flow.engine.bamboo.scene.spider.append_deploy_ctl_flow import AppendDeployCTLFlow
from backend.flow.engine.bamboo.scene.spider.db_table_backup import TenDBClusterDBTableBackupFlow
from backend.flow.engine.bamboo.scene.spider.full_backup import TenDBClusterFullBackupFlow
from backend.flow.engine.bamboo.scene.spider.import_sqlfile_flow import ImportSQLFlow
from backend.flow.engine.bamboo.scene.spider.remote_local_slave_recover import TenDBRemoteSlaveLocalRecoverFlow
from backend.flow.engine.bamboo.scene.spider.remote_master_fail_over import RemoteMasterFailOverFlow
from backend.flow.engine.bamboo.scene.spider.remote_master_slave_migrate import TendbClusterMigrateRemoteFlow
from backend.flow.engine.bamboo.scene.spider.remote_master_slave_swtich import RemoteMasterSlaveSwitchFlow
from backend.flow.engine.bamboo.scene.spider.remote_slave_recover import TenDBRemoteSlaveRecoverFlow
from backend.flow.engine.bamboo.scene.spider.spider_add_mnt import TenDBClusterAddSpiderMNTFlow
from backend.flow.engine.bamboo.scene.spider.spider_add_nodes import TenDBClusterAddNodesFlow
from backend.flow.engine.bamboo.scene.spider.spider_checksum import SpiderChecksumFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_deploy import TenDBClusterApplyFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_destroy import TenDBClusterDestroyFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_disable_deploy import SpiderClusterDisableFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_enable_deploy import SpiderClusterEnableFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_flashback import TenDBClusterFlashbackFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_metadata_import_flow import SpiderClusterMetadataImportFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_rollback_flow import TenDBRollBackDataFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_standardize_flow import SpiderClusterStandardizeFlow
from backend.flow.engine.bamboo.scene.spider.spider_cluster_truncate_database import SpiderTruncateDatabaseFlow
from backend.flow.engine.bamboo.scene.spider.spider_partition import SpiderPartitionFlow
from backend.flow.engine.bamboo.scene.spider.spider_partition_cron import SpiderPartitionCronFlow
from backend.flow.engine.bamboo.scene.spider.spider_reduce_mnt import TenDBClusterReduceMNTFlow
from backend.flow.engine.bamboo.scene.spider.spider_reduce_nodes import TenDBClusterReduceNodesFlow
from backend.flow.engine.bamboo.scene.spider.spider_remotedb_rebalance_flow import TenDBRemoteRebalanceFlow
from backend.flow.engine.bamboo.scene.spider.spider_rename_database_flow import SpiderRenameDatabaseFlow
from backend.flow.engine.bamboo.scene.spider.spider_slave_cluster_deploy import TenDBSlaveClusterApplyFlow
from backend.flow.engine.bamboo.scene.spider.spider_slave_cluster_destroy import TenDBSlaveClusterDestroyFlow
from backend.flow.engine.controller.base import BaseController


class SpiderController(BaseController):
    """
    spider相关调用
    """

    def spider_cluster_apply_scene(self):
        """
        部署tenDB cluster(spider cluster) 部署场景
        """
        flow = TenDBClusterApplyFlow(root_id=self.root_id, data=self.ticket_data)
        flow.deploy_cluster()

    def spider_cluster_apply_no_slave(self):
        """
        部署tenDB cluster(spider cluster) 部署场景，没有remote slave 单节点中控节点
        """
        flow = TenDBClusterApplyFlow(root_id=self.root_id, data=self.ticket_data)
        flow.deploy_cluster_no_slave()

    def spider_cluster_destroy_scene(self):
        """
        下架tenDB cluster(spider cluster) 部署场景
        """
        flow = TenDBClusterDestroyFlow(root_id=self.root_id, data=self.ticket_data)
        flow.destroy_cluster()

    def add_spider_mnt_scene(self):
        flow = TenDBClusterAddSpiderMNTFlow(root_id=self.root_id, data=self.ticket_data)
        flow.add_spider_mnt()

    def spider_checksum(self):
        """
        mysql 数据校验
        """
        flow = SpiderChecksumFlow(root_id=self.root_id, data=self.ticket_data)
        flow.spider_checksum_flow()

    def spider_partition(self):
        """
        spider 表分区
        """
        flow = SpiderPartitionFlow(root_id=self.root_id, data=self.ticket_data)
        flow.spider_partition_flow()

    def spider_partition_cron(self):
        """
        spider 表分区定时任务
        """
        flow = SpiderPartitionCronFlow(root_id=self.root_id, data=self.ticket_data)
        flow.spider_partition_cron_flow()

    def spider_cluster_disable_scene(self):
        """
        禁用spider集群场景
        """
        flow = SpiderClusterDisableFlow(root_id=self.root_id, data=self.ticket_data)
        flow.disable_spider_cluster_flow()

    def spider_cluster_enable_scene(self):
        """
        启用spider集群场景
        """
        flow = SpiderClusterEnableFlow(root_id=self.root_id, data=self.ticket_data)
        flow.enable_spider_cluster_flow()

    def spider_slave_cluster_apply_scene(self):
        """
        调用tenDB slave cluster(spider slave cluster) 部署场景
        """
        flow = TenDBSlaveClusterApplyFlow(root_id=self.root_id, data=self.ticket_data)
        flow.deploy_slave_cluster()

    def rename_database(self):
        flow = SpiderRenameDatabaseFlow(
            root_id=self.root_id, data=self.ticket_data, cluster_type=ClusterType.TenDBCluster.value
        )
        flow.rename_database()

    def spider_semantic_check_scene(self):
        """
        spider语义测试场景
        """
        flow = ImportSQLFlow(root_id=self.root_id, data=self.ticket_data)
        flow.sql_semantic_check_flow()

    def spider_sql_import_scene(self):
        flow = ImportSQLFlow(root_id=self.root_id, data=self.ticket_data)
        flow.import_sqlfile_flow()

    def truncate_database(self):
        flow = SpiderTruncateDatabaseFlow(
            root_id=self.root_id, data=self.ticket_data, cluster_type=ClusterType.TenDBCluster.value
        )
        flow.truncate_database()

    def database_table_backup(self):
        flow = TenDBClusterDBTableBackupFlow(root_id=self.root_id, data=self.ticket_data)
        flow.backup_flow()

    def add_spider_nodes_scene(self):
        """
        扩容接入层的场景
        """
        flow = TenDBClusterAddNodesFlow(root_id=self.root_id, data=self.ticket_data)
        flow.add_spider_nodes()

    def full_backup(self):
        flow = TenDBClusterFullBackupFlow(root_id=self.root_id, data=self.ticket_data)
        flow.full_backup_flow()

    def reduce_spider_nodes_scene(self):
        """
        缩容接入层的场景
        """
        flow = TenDBClusterReduceNodesFlow(root_id=self.root_id, data=self.ticket_data)
        flow.reduce_spider_nodes()

    def flashback(self):
        flow = TenDBClusterFlashbackFlow(
            root_id=self.root_id, data=self.ticket_data, cluster_type=ClusterType.TenDBCluster.value
        )
        flow.flashback()

    def tendb_cluster_remote_switch_scene(self):
        """
        remote端互切的场景
        """
        flow = RemoteMasterSlaveSwitchFlow(root_id=self.root_id, data=self.ticket_data)
        flow.remote_switch()

    def tendbcluster_remote_fail_over_scene(self):
        """
        remote端主故障切换的场景
        """
        flow = RemoteMasterFailOverFlow(root_id=self.root_id, data=self.ticket_data)
        flow.remote_fail_over()

    def tendb_cluster_remote_rebalance(self):
        """
        remote 节点扩缩容同步数据(重均衡)
        """
        flow = TenDBRemoteRebalanceFlow(root_id=self.root_id, ticket_data=self.ticket_data)
        flow.tendb_migrate()

    def tendb_cluster_remote_migrate(self):
        """
        remote 节点主从成对迁移
        """
        flow = TendbClusterMigrateRemoteFlow(root_id=self.root_id, ticket_data=self.ticket_data)
        flow.migrate_master_slave_flow()

    def tendb_cluster_remote_slave_recover(self):
        """
        remote 远程slave节点恢复
        """
        flow = TenDBRemoteSlaveRecoverFlow(root_id=self.root_id, ticket_data=self.ticket_data)
        flow.tendb_remote_slave_recover()

    def tendb_cluster_remote_local_recover(self):
        """
        remote 本地恢复
        """
        flow = TenDBRemoteSlaveLocalRecoverFlow(root_id=self.root_id, ticket_data=self.ticket_data)
        flow.tendb_remote_slave_local_recover()

    def tendb_cluster_rollback_data(self):
        """
        tendb cluster 定点回档
        """
        flow = TenDBRollBackDataFlow(root_id=self.root_id, ticket_data=self.ticket_data)
        flow.tendb_rollback_data()

    def destroy_tendb_slave_cluster(self):
        """
        tendb cluster 只读接入层下架
        """
        flow = TenDBSlaveClusterDestroyFlow(root_id=self.root_id, data=self.ticket_data)
        flow.destroy_slave_cluster()

    def reduce_spider_mnt_scene(self):
        """
        tendb cluster 运维节点下架
        """
        flow = TenDBClusterReduceMNTFlow(root_id=self.root_id, data=self.ticket_data)
        flow.reduce_spider_mnt()

    def append_deploy_ctl_scene(self):
        """
        转移spider cluster 从Gcs到dbm 系统需要的操作
        """
        flow = AppendDeployCTLFlow(root_id=self.root_id, data=self.ticket_data)
        flow.run()

    def tendbcluster_standardize_scene(self):
        flow = SpiderClusterStandardizeFlow(root_id=self.root_id, data=self.ticket_data)
        flow.standardize()

    def metadata_import_scene(self):
        flow = SpiderClusterMetadataImportFlow(root_id=self.root_id, data=self.ticket_data)
        flow.import_meta()

File: dbm-ui/frontend/src/services/source/spider.ts
/*
 * TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
 *
 * Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
 * the specific language governing permissions and limitations under the License.
 */

import SpiderModel from '@services/model/spider/spider';
import SpiderMachineModel from '@services/model/spider/spiderMachine';
import TendbClusterModel from '@services/model/spider/tendbCluster';
import TendbInstanceModel from '@services/model/spider/tendbInstance';

import http from '../http';
import type { ListBase, ResourceTopo } from '../types';

const getRootPath = () => `/apis/mysql/bizs/${window.PROJECT_CONFIG.BIZ_ID}/spider_resources`;

/**
 * 获取 TendbCluster 集群列表
 */
export function getTendbClusterList(params: Record<string, any> = {}) {
  return http.get<ListBase<TendbClusterModel[]>>(`${getRootPath()}/`, params).then((data) => ({
    ...data,
    results: data.results.map(
      (item) =>
        new TendbClusterModel(
          Object.assign(item, {
            permission: Object.assign({}, item.permission, data.permission),
          }),
        ),
    ),
  }));
}

/**
 * 获取 spider 集群列表
 */
export function getSpiderList(params: {
  limit?: number;
  offset?: number;
  id?: number;
  name?: string;
  ip?: string;
  domain?: string;
  exact_domain?: string;
  creator?: string;
  version?: string;
  cluster_ids?: string | number;
  db_module_id?: number;
  region?: string;
}) {
  return http.get<ListBase<SpiderModel[]>>(`${getRootPath()}/`, params).then((data) => ({
    ...data,
    results: data.results.map((item: SpiderModel) => new SpiderModel(item)),
  }));
}

/**
 * 根据业务 id 获取 spider 集群列表
 */
export function getSpiderListByBizId(params: {
  bk_biz_id: number;
  limit?: number;
  offset?: number;
  id?: number;
  name?: string;
  ip?: string;
  domain?: string;
  creator?: string;
  version?: string;
  cluster_ids?: number[];
  db_module_id?: number;
  region?: string;
}) {
  return http
    .get<ListBase<SpiderModel[]>>(`/apis/mysql/bizs/${params.bk_biz_id}/spider_resources/`, params)
    .then((data) => ({
      ...data,
      results: data.results.map((item: SpiderModel) => new SpiderModel(item)),
    }));
}

/**
 * 查询资源列表
 */
export function getResources(params: {
  bk_biz_id: number;
  limit: number;
  offset: number;
  type: string;
  cluster_ids?: number[] | number;
  dbType: string;
}) {
  return http.get<ListBase<TendbClusterModel[]>>(`${getRootPath()}/`, params);
}

/**
 * 根据业务 ID 查询资源列表
 */
export function getResourcesByBizId(params: {
  bk_biz_id: number;
  limit: number;
  offset: number;
  type: string;
  cluster_ids?: number[] | number;
  dbType: string;
}) {
  return http.get<ListBase<TendbClusterModel[]>>(`/apis/mysql/bizs/${params.bk_biz_id}/spider_resources/`, params);
}

/**
 * 查询表格信息
 */
export function getSpiderTableFields() {
  return http.get<
    {
      key: string;
      name: string;
    }[]
  >(`${getRootPath()}/get_table_fields/`);
}

/**
 * 获取集群实例列表
 */
export function getSpiderInstanceList(params: Record<string, any>) {
  return http.get<ListBase<TendbInstanceModel[]>>(`${getRootPath()}/list_instances/`, params).then((res) => ({
    ...res,
    results: res.results.map((data) => new TendbInstanceModel(data)),
  }));
}

/**
 * 获取集群详情
 */
export function getSpiderDetail(params: { id: number }) {
  return http.get<SpiderModel>(`${getRootPath()}/${params.id}/`).then((data) => new SpiderModel(data));
}

/**
 * 获取集群拓扑
 */
export function getSpiderTopoGraph(params: { cluster_id: number }) {
  return http.get<ResourceTopo>(`${getRootPath()}/${params.cluster_id}/get_topo_graph/`);
}

/**
 * 导出集群数据为 excel 文件
 */
export function exportSpiderClusterToExcel(params: { cluster_ids?: number[] }) {
  return http.post<string>(`${getRootPath()}/export_cluster/`, params, { responseType: 'blob' });
}

/**
 * 导出实例数据为 excel 文件
 */
export function exportSpiderInstanceToExcel(params: { bk_host_ids?: number[] }) {
  return http.post<string>(`${getRootPath()}/export_instance/`, params, { responseType: 'blob' });
}

/**
 * 查询主机列表
 */
export function getSpiderMachineList(params: {
  limit?: number;
  offset?: number;
  bk_host_id?: number;
  ip?: string;
  machine_type?: string;
  bk_os_name?: string;
  bk_cloud_id?: number;
  bk_agent_id?: string;
  instance_role?: string;
  spider_role?: string;
  creator?: string;
}) {
  return http.get<ListBase<SpiderMachineModel[]>>(`${getRootPath()}/list_machines/`, params).then((data) => ({
    ...data,
    results: data.results.map((item) => new SpiderMachineModel(item)),
  }));
}

/**
 * 获取 spider 集群详情
 * @param id 集群 ID
 */
export const getSpiderDetails = (params: { id: number }) =>
  http.get<TendbClusterModel>(`${getRootPath()}/${params.id}/`);

/**
 * 获取 spider 实例详情
 */
export const getSpiderInstanceDetails = (params: { instance_address: string; cluster_id: number }) =>
  http.get<TendbInstanceModel>(`${getRootPath()}/retrieve_instance/`, params);

export const getList = function (params: Record<string, any>) {
  return http.get<ListBase<SpiderModel[]>>(`${getRootPath()}/`, params).then((data) => ({
    ...data,
    results: data.results.map((item: SpiderModel) => new SpiderModel(item)),
  }));
};

export const getDetail = function (params: { id: number }) {
  return http.get<SpiderModel>(`${getRootPath()}/${params.id}/`).then((data) => new SpiderModel(data));
};
